cmd/paramgen/internal/paramgen.go

```diff
@@ -0,0 +1,628 @@
+	if err != nil {
+		return nil, err
+	}
+
+	// now find requested type in that package
+	ts, file, err := p.findType(pkg, se.Sel.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// back up current file and replace it because we are now working with
+	// another file, we want to revert this once we are done parsing this type
+	backupFile := p.file
+	backupPkg := p.pkg
+	p.file = file
+	p.pkg = pkg
+	defer func() {
+		p.file = backupFile
+		p.pkg = backupPkg
+	}()
+
+	return p.parseTypeSpec(ts)
+}
+
+func (p *paramsParser) findPackage(importPath string) (*ast.Package, error) {
+	// first cleanup string
+	importPath = strings.Trim(importPath, `"`)
+
+	if !strings.HasPrefix(importPath, p.mod.Path) {
+		// we only allow types declared in the same module
+		return nil, fmt.Errorf("we do not support parameters from package %v (please use builtin types or time.Duration)", importPath)
+	}
+
+	if pkg, ok := p.imports[importPath]; ok {
+		// it's cached already
+		return pkg, nil
+	}
+
+	pkgDir := p.mod.Dir + strings.TrimPrefix(importPath, p.mod.Path)
+	pkg, err := parsePackage(pkgDir)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse package dir %q: %w", pkgDir, err)
+	}
+
+	// cache it for future use
+	p.imports[importPath] = pkg
+	return pkg, nil
+}
+
+func (p *paramsParser) findType(pkg *ast.Package, typeName string) (*ast.TypeSpec, *ast.File, error) {
+	var file *ast.File
+	var found *ast.TypeSpec
+	for _, f := range pkg.Files {
+		ast.Inspect(f, func(node ast.Node) bool {
+			ts, ok := node.(*ast.TypeSpec)
+			if !ok {
+				return true
+			}
+			if ts.Name.Name != typeName {
+				return true
+			}
+
+			// found our type, store the file and type
+			file = f
+			found = ts
+			return false
+		})
+		if found != nil {
+			// already found the type
+			break
+		}
+	}
+	if found == nil {
+		return nil, nil, fmt.Errorf("could not find type %v in package %v", typeName, pkg.Name)
+	}
+	return found, file, nil
+}
+
+func (p *paramsParser) findImportSpec(se *ast.SelectorExpr) (*ast.ImportSpec, error) {
+	impName := se.X.(*ast.Ident).Name
+	for _, i := range p.file.Imports {
+		if (i.Name != nil && i.Name.Name == impName) ||
+			strings.HasSuffix(strings.Trim(i.Path.Value, `"`), impName) {
+			return i, nil
+		}
+	}
+	return nil, fmt.Errorf("could not find import %q", impName)
+}
+
+func (p *paramsParser) attachPrefix(f *ast.Field, params map[string]sdk.Parameter) map[string]sdk.Parameter {
+	// attach prefix if a tag is present or if the field is named
+	prefix := p.getTag(f.Tag, TagParamName)
+	if prefix == "" && len(f.Names) > 0 {
+		prefix = p.formatFieldName(f.Names[0].Name)
+	}
+	// if it's a struct, then prefix is the struct name
+	if n, ok := f.Type.(*ast.Ident); ok && prefix == "" {
+		prefix = p.formatFieldName(n.Name)
+	}
+	if prefix == "" {
+		// no prefix to attach
+		return params
+	}
+
+	prefixedParams := make(map[string]sdk.Parameter)
+	for k, v := range params {
+		prefixedParams[prefix+"."+k] = v
```

Comment:
Since we have separators in constants it would be nice to have the dot in a constant as well.